geomesa-plugin/src/main/scala/org/locationtech/geomesa/plugin/security/UserNameRoles.scala

```diff
@@ -0,0 +1,113 @@
+  // this must be created on the fly with the user, and since standard (XML)
+  // RoleServices don't write back to their backing configuration
+  // we need everything to happen as a layer on top of the configured answers
+  abstract override def getRolesForUser(username: String): jSortedSet[GeoServerRole] = {
+    import collection.JavaConversions._
+    val roles = new jTreeSet[GeoServerRole]()
+    roles.addAll(super.getRolesForUser(username))
+    roles.add(storedUserRoles(username))
+    println(s"Retrieved roles for $username: ${roles.map(_.toString).mkString(",")}")
+    Collections.unmodifiableSortedSet(roles)
+  }
+
+  // the rest of this is to try to provide consistency with the above piece
+
+  // the only group for a UserRole is the user's identity group
+  abstract override def getGroupNamesForRole(role: GeoServerRole): jSortedSet[String] = {
+    if (isUserRole(role)) {
+      val groupNames = new jTreeSet[String]()
+      groupNames.add(role.getUserName)
+      Collections.unmodifiableSortedSet(groupNames)
+    } else super.getGroupNamesForRole(role)
+  }
+
+  // the only user name in a UserRole is the user itself
+  abstract override def getUserNamesForRole(role: GeoServerRole): jSortedSet[String] = {
+    if (isUserRole(role)) {
+      val userNames = new jTreeSet[String]()
+      userNames.add(role.getUserName)
+      Collections.unmodifiableSortedSet(userNames)
+    } else super.getUserNamesForRole(role)
+  }
+
+  abstract override def getParentMappings: jMap[String, String] = {
+    val mappings = new jHashMap[String, String]
+    mappings.putAll(super.getParentMappings)
+    for ((username, _) <- storedUserRoles) mappings.put(username, null)
+    Collections.unmodifiableMap(mappings)
+  }
+
+  // if we're accessing by group it's almost certainly NOT a user's identity group (I hope)
+  // thus getRolesForGroup(groupname: String): jSortedSet[GeoServerRole] stays the same
+
+  // can't see how to automagically access the bean-provided UserService, so can't list all
+  // user name roles in the call to getRoles(). However, we CAN list the roles we've seen before...
+  abstract override def getRoles: jSortedSet[GeoServerRole] = {
+    val roles = new jTreeSet[GeoServerRole]
+    roles.addAll(super.getRoles)
+    for (ugService <- getSecurityManager.loadUserGroupServices; user <- ugService.getUsers) {
+      val username = user.getUsername
+      val userrole = storedUserRoles(username)
+      roles.add(userrole)
+    }
+    Collections.unmodifiableSortedSet(roles)
+  }
+}
+
+class XMLUserNameRoleService extends XMLRoleService with UserNameRoles
+
+// The RoleService seems to be deployed from the SecurityProvider, so this will replace
+// that with one that creates an XMLUserNameRoleService instead of an XMLRoleService
+// In turn, the SecurityProvider is deployed as a bean in the gs-main JAR, specified in
+// applicationSecurityContext.xml.  To replace it
+// 1) unzip the gs-main JAR into a spare directory
```

Comment:
is there a way to do this with SPI or something...I don't believe we can go around zipping/unzipping jar files in production easily.